map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,162 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+uint64_t constexpr kMaxExtrapolationTimeMs = 1000;
+uint64_t constexpr kExtrapolationPeriodMs = 200;
+double constexpr kMaxExtrapolationSpeedMPerS = 120.0;
+
+double LinearExtrapolationOfOneParam(double param1, double param2, uint64_t timeBetweenPointsMs,
```

Comment:
I'm not sure about the interface: why not pass timeAt as the last parameter, i.e. the absolute timestamp?

Reply:
Are you talking about something line that:
```c++
double LinearExtrapolationOfOneParam(double param1, uint64_t timestamp1, double param2, uint64_t timestamp2, uint64_t timeBetweenPointsMs, uint64_t timeAfterParam2Ms)
{
  return (param2 - param1) * timeAfterParam2Ms / (time2 - time1);
}
```
?

Reply:
No, what I meant is ```time3``` instead of ```timeAfter2```, i.e. ```t3``` instead of ```t3-t2```.

Reply:
I'd prefer to calculate:
```c++
  auto const timeBetweenPointsMs =
      static_cast<uint64_t>((gpsInfo2.m_timestamp - gpsInfo1.m_timestamp) * 1000);
```
only once.

Reply:
I wrote LinearExtrapolator instead.

Reply:
> No, what I meant is time3 instead of timeAfter2, i.e. t3 instead of t3-t2

Are you sure it'll be better? To do this I need to add some conversion from double to uint64_t at the place of call:
``` 
         location::GpsInfo gpsInfo =
              LinearExtrapolation(m_beforeLastGpsInfo, m_lastGpsInfo, extrapolationTimeMs);
```